File: src/networking/RicartArgawala.java
package networking;

import java.io.IOException;
import java.util.Collections;
import java.util.List;
import java.util.Random;
import java.util.Set;
import java.util.TreeSet;
import java.util.Vector;

import org.apache.xmlrpc.XmlRpcException;

import networking.Client.RemoteNode;

public class RicartArgawala extends SyncAlgorithm implements Runnable {
	
	private volatile Set<String> requestsQueue;
	private volatile Set<String> okayList;
	
	private volatile boolean canAccess;
	private long timestamp;
	private volatile boolean isSyncing;
	
	public RicartArgawala(List<RemoteNode> networkInput, String ip)
	{
		// pay attention to the difference between network and this.network
		super(networkInput, ip);
	    Random randomGenerator = new Random();
		timestamp = randomGenerator.nextInt(50);
		requestsQueue = new TreeSet<String>();
		okayList      = Collections.synchronizedSet( new TreeSet<String>() );
		isPending     = false;
		canAccess     = false;
		isSyncing     = false;
		instance      = this;
	}
	
	public void requestReceived( String ip, long timestamp )
	{
//		System.out.println("Received request from ip "+ip+" timestamp = "+timestamp+" own timestamp is "+this.timestamp);
		boolean sendOk = false;
		boolean addToQueue = false;
		
		boolean calcIsDone = isCalcDone();
		boolean isPending = isPending();

		if ( calcIsDone && !isPending )
		{
			sendOk = true;
		}
		else if ( !calcIsDone )
		{
			addToQueue = true;
		}
		else if ( isPending || isSyncing )
		{
			if ( ( timestamp == this.timestamp && ip.compareTo(this.ip) > 0 ) || 
					   timestamp < this.timestamp )
			{
				sendOk = true;
			}
			else
			{
				addToQueue = true;
			}
		}

		if ( sendOk )
		{
			sendOk(ip);
		}
		else
		{
			synchronized (this.requestsQueue)
			{
				requestsQueue.add(ip);
			}
		}



		
//		synchronized (this.requestsQueue) {
//			if ( isCalcDone() && !isPending() )
//			{
////					System.out.println("Case isCalcDone() && !isPending()");
//					// send OK
//					
//					sendOk(ip);
//			}
//			else if ( !isCalcDone() )
//			{
////				System.out.println("!isCalcDone");
//				requestsQueue.add(ip);
//			}
//			else if ( isPending() )
//			{
//				// queue request
//				if ( ( timestamp == this.timestamp && ip.compareTo(this.ip) > 0 ) || 
//					   timestamp < this.timestamp )
//				{
////					System.out.println("case timestamp is smaller");
//						sendOk(ip);
//				}
//				else
//				{
////					System.out.println("case timestamp is bigger");
//					requestsQueue.add(ip);
//				}
//			}
//		}
	}

	private void sendOk(String ip) {
//		System.out.println("Sending okay to node "+ip+" timestamp = "+timestamp);
		Vector<String> params = new Vector<String>();
		params.add(this.ip);

		for ( RemoteNode node : this.network )
		{
			if ( node.ip.compareTo(ip) == 0 )
			{
				try {
					node.rpc.execute("RicartArgawalaAux.okReceived", params);
				} catch (XmlRpcException | IOException e) {
				
					e.printStackTrace();
				}
				break;
			}
		}
	}

	public synchronized void okReceived(String ip) {
//		System.out.println("==>Received okay from "+ip);
		okayList.add(ip);
//		System.out.println("  okayList size is "+okayList.size());
	}

	@Override
	public void run() {

		boolean isSyncDone = false;
		while (!canStart());
		
		while ( !isSessionDone() )
		{
			isSyncDone = false;
			boolean isPending;
			synchronized ( this.mutualLock ) {
				isPending = isPending();
			}
			
			if ( isPending )
			{
				isSyncing = true;
//						System.out.println("Pending request detected\n");
				// request from all nodes
				okayList.clear();
				broadcastRequest();
				
				// wait for okay from all
				while( okayList.size() < ( network.size() - 1 ) ); // -1 because of self node
				isSyncDone = true;
				setAccess( true );
			}
				
				
				
			if ( isSyncDone )
			{
				System.out.println("====> CS ra enter");
				
					// can access now
					while (isPending());
					while (!isCalcDone());

					setAccess( false );
					// send okay to all processes in queue
					synchronized ( this.requestsQueue) {
						sendOkayToQueueNodes();
					}

					timestamp++;
				
				System.out.println("<==== CS ra exit");
				isSyncing = false;
			}
		}
	}

	private synchronized boolean isPending() {
		return isPending;
	}

	private synchronized void sendOkayToQueueNodes() {
		
		Vector<String> params = new Vector<String>();

		params.add(this.ip);
		
		for (String node : requestsQueue )
		{
			for ( RemoteNode rNode : network )
			{
				if (node.compareTo(rNode.ip) == 0)
				{
//					System.out.println("Sending okay to queue node "+rNode.ip);
					try {
						rNode.rpc.execute("RicartArgawalaAux.okReceived", params);
					} catch (XmlRpcException | IOException e) {
						e.printStackTrace();
					}
				}				
			}
		}

	}

	private void broadcastRequest() {
		System.out.println("Broadcasting request:");
		Vector<String> params = new Vector<String>();
		
		params.add(this.ip);
		params.add(Long.toString(this.timestamp));
		
		for (RemoteNode node : network )
		{
			if (node.ip.compareTo(this.ip) != 0 )
			{
				try {
					node.rpc.execute("RicartArgawalaAux.requestReceived", params);
				} catch (XmlRpcException | IOException e) {
					e.printStackTrace();
				}
			}
		}
	}

	@Override
	public synchronized boolean canAccess() {
		return canAccess;
	}
	
	private synchronized void setAccess(boolean accessValue)
	{
		canAccess = accessValue;
	}
}

File: CSharp/RicartArgawala.cs
using System;
using System.Collections.Generic;
using CookComputing.XmlRpc;
using System.Linq;
using System.Text;

namespace Networking
{
    class RicartArgawala : SyncAlgorithm
        {

            private volatile HashSet<string> requestsQueue;
            private volatile HashSet<string> okayList;

            private volatile bool canAccess_Renamed;
            private volatile bool isSyncing;

            private long timestamp;
            bool _keepRunning = true;

            public RicartArgawala(IList<RemoteNode> network, string ip) : base(network, ip)
            {
                // pay attention to the difference between network and this.network
                Random randomGenerator = new Random();
                timestamp = randomGenerator.Next(50);
               // Console.WriteLine("timestamp init to: " + timestamp);

                requestsQueue = new HashSet<string>();
                okayList = new HashSet<string>();
                isPending = false;
                canAccess_Renamed = false;
                isSyncing = false;
                instance = this;
            }

            public virtual void requestReceived(string ip, long timestamp)
            {
                // Console.WriteLine("Received request from ip " + ip + " timestamp = " + timestamp + " mystamp = " + this.timestamp);
                bool sendOk_r = false;
                bool addToQueue = false;

                bool calcIsDone = CalcDone;
                bool isPending = Pending;

                if (calcIsDone && !isPending)
                    {
                        // Console.WriteLine("CalcDone && !Pending");
                        // send OK
                        sendOk_r = true;
                    }
                    else if (!calcIsDone)
                    {
                        addToQueue = true;
                    }
                    else if (isPending || _keepRunning)
                    {
                        if ((timestamp == this.timestamp && ip.CompareTo(this.ip) > 0) ||
                               timestamp < this.timestamp)
                        {
                            sendOk_r = true;
                        }
                        else
                        {
                            addToQueue = true;
                        }
                    }
                    if (sendOk_r)
                    {
                        sendOk(ip);
                    }
                    else
                    {
                        lock (this.requestsQueue)
                        {
                            requestsQueue.Add(ip);
                        }
                    }
                }
        

       private void sendOk(string ip)
  		{
            //Console.WriteLine("Sending okay to node " + ip + " timestamp = " + timestamp );
            
            NetworkClientInterface executer = XmlRpcProxyGen.Create<NetworkClientInterface>();
            executer.AttachLogger(new XmlRpcDebugLogger());

            executer.Url = generate_url(ip);

            foreach (RemoteNode node in this.network)
            {
                if (node.getIP().CompareTo(ip) == 0)
                {
                    executer.okReceived(this.ip);
                    break;
                }
            }

			
		}

            public virtual void okReceived(string ip)
            {
                lock (this)
                {
                   // Console.WriteLine("==>Received okay from " + ip);
                    okayList.Add(ip);
                    //Console.WriteLine("   okayList size is" + okayList.Count);
                }
            }

            public void RequestStop()
            {
                _keepRunning = false;
            }

            public void run()
            {//192.168.0.33

                bool isSyncDone = false;

                Console.WriteLine("Thread  RicartArgwala is running");
                while (!getcanStart()) ;

                while (_keepRunning)
                {
                    isSyncDone = false;
                    bool isPending;
			        lock ( this.mutualLock ) {
				        isPending = Pending;
			        }

                    if (isPending)
                    {
                        isSyncing = true;
                     //   Console.WriteLine("Pending request detected\n");
                        // request from all nodes
                        okayList.Clear();
                        broadcastRequest();
                       
                        // wait for okay from all                        
                        while ((okayList.Count < (network.Count - 1))&&(_keepRunning)); // -1 because of self node
                        isSyncDone = true; 
                        Access = true;
                        }

                    if (isSyncDone)
                    {
                        Console.WriteLine("====> CS ra enter");

                        // can access now
                        while (Pending);
                        while (!CalcDone);

                        Access = false;

                        // send okay to all processes in queue
                        lock (this.requestsQueue)
                        {
                            sendOkayToQueueNodes();
                        }

                        timestamp++;

                        Console.WriteLine("<==== CS exit");

                        isPending = false;
                        
                    }
                }
             Console.WriteLine("\nThread  RicartArgwala is terminated\n");
            }


            public String generate_url(String ip)
            {
                return "http://" + ip + ":" + 5000 + "/RPC2";
            }


       private void sendOkayToQueueNodes(){

			foreach (string node_ip in requestsQueue)
  			   {
                //Console.WriteLine("Sending okay to queue node " + node_ip);
                foreach(RemoteNode rNode in network){
                   if (node_ip.CompareTo(rNode.getIP()) == 0)
                   {
                       NetworkClientInterface executer = XmlRpcProxyGen.Create<NetworkClientInterface>();
                       executer.AttachLogger(new XmlRpcDebugLogger());

                       executer.Url = generate_url(node_ip);
                       executer.okReceived(this.ip);
                   }
                }
                }

		}

       private void broadcastRequest()
		{
			Console.WriteLine("Boradcasting request:");
			foreach (RemoteNode node in network)
            {
                
				if (node.getIP().CompareTo(this.ip) != 0)
                {
                    //Console.WriteLine("Boradcasting request to:" + node.getIP());
                    NetworkClientInterface executer = XmlRpcProxyGen.Create<NetworkClientInterface>();
                    executer.AttachLogger(new XmlRpcDebugLogger());

                    executer.Url = node.getURL();
                    executer.requestReceived(this.ip, Convert.ToString(this.timestamp));

				}
			}
		}



       private bool Pending
       {
           get
           {
               lock (this)
               {
                   return isPending;
               }
           }
       }

        
            public override bool canAccess()
            {
                lock (this)
                {
                    return canAccess_Renamed;
                }
            }

            private bool Access
            {
                set
                {
                    lock (this)
                    {
                        canAccess_Renamed = value;
                    }
                }
            }

        }





    }
